addon/-private/properties/collection.js

```diff
@@ -1,92 +1,177 @@
-    collection[Symbol.iterator] = iteratorMethod;
-  }
+    if (!_items[index]) {
+      let {
+        _itemDefinition,
+        _collectionScope,
+        _itemScope,
+        _testContainer,
+        _resetScope,
+        _parentNode,
+        _key
+      } = this;
 
-  // Change the key of the root node
-  Ceibo.meta(collection).key = `${key}()`;
+      let filters = merge({}, { scope: _collectionScope, at: index });
+      let scope = buildSelector({}, _itemScope, filters);
 
-  return collection;
-}
+      let finalizedItemDefinition = merge(
+        {
+          testContainer: _testContainer
+        },
+        _itemDefinition,
+        {
+          scope,
+          resetScope: _resetScope
+        }
+      );
 
-function generateItem(node, index, definition, key) {
-  let filters = merge({}, { scope: definition.scope, at: index });
-  let scope = buildSelector({}, definition.itemScope, filters);
-
-  let tree = create(
-    merge(
-      {
-        testContainer: definition.testContainer
-      },
-      definition.item,
-      {
-        scope,
-        resetScope: definition.resetScope
-      }
-    ), { parent: node });
+      let tree = create(finalizedItemDefinition, { parent: _parentNode });
 
-  // Change the key of the root node
-  Ceibo.meta(tree).key = `${key}(${index})`;
+      // Change the key of the root node
+      Ceibo.meta(tree).key = `${_key}(${index})`;
 
-  return tree;
-}
+      _items[index] = tree;
+    }
+
+    return _items[index];
+  }
+
+  eq(index) {
+    return this._getItem(index);
+  }
+
+  toArray() {
+    let { length } = this;
 
-function toArrayMethod(node, definition, key) {
-  return function() {
     let array = Ember.A();
-    let index;
-    let count;
 
-    for (index = 0, count = this.count; index < count; index++) {
-      array.push(generateItem(node, index, definition, key));
+    for (let i = 0; i < length; i++) {
+      array.push(this.eq(i));
     }
 
     return array;
-  };
-}
+  }
 
-function delegateToArray(enumerable, method) {
-  if (typeof (enumerable[method]) === 'undefined') {
-    enumerable[method] = function(...args) {
-      return this.toArray()[method](...args);
-    };
+  map(...args) {
+    return this.toArray().map(...args);
+  }
+
+  mapBy(...args) {
+    return this.toArray().mapBy(...args);
+  }
+
+  filter(...args) {
+    return this.toArray().filter(...args);
+  }
+
+  filterBy(...args) {
+    return this.toArray().filterBy(...args);
+  }
+
+  forEach(...args) {
+    return this.toArray().forEach(...args);
+  }
+
+  findOne(query) {
+    let result = this.findAll(query);
+
+    assert(`Expected at most one result from findOne query, but found ${result.length}`, result.length <= 1);
+
+    return result[0];
+  }
+
+  findAll(query) {
+    let predicate;
+
+    if (typeof (query) === 'object') {
+      predicate = (item) => {
+        let isMatch = true;
+
+        for (let key in query) {
+          isMatch = isMatch && item[key] === query[key];
```

Comment:
I wonder if we need to support array comparisons here. 

Suppose some parking lot app allows attendants to enter a list of driver names, each of which is followed by a series of inputs used to enter information about the driver's car:
```html
<div data-test-driver>
  <input type="text" placeholder="Driver Name">
  <input type="text" placeholder="Car Make">
  <input type="text" placeholder="Car Model">
  <input type="text" placeholder="Car License Plate">
</div>
```
```javascript
// in the page object . . .
drivers: collection({
  scope: '[data-test-driver]',
  driverInfo: text('input', { multiple: true })
})
```
```javascript
// in some test . . .
const driver = page.drivers.findOne({ 
  driverInfo: ['Grace Hopper', 'Nissan', 'Sentra', 'FOOBAR'] 
});
```

Now, this would make a lot more sense:

```html
<div data-test-driver>
  <input type="text" data-test-driver-name placeholder="Driver Name">
  <input type="text" data-test-car-make placeholder="Car Make">
  <input type="text" data-test-car-model placeholder="Car Model">
  <input type="text" data-test-car-plate placeholder="Car License Plate">
</div>
```
```javascript
// in the page object . . .
drivers: collection({
  scope: '[data-test-driver]',
  driverName: text('[data-test-driver-name]'),
  carMake: text('[data-test-car-make]'),
  carModel: text('[data-test-car-model]'),
  carPlate: text('[data-test-car-plate]')
})
```
```javascript
// in some test . . .
const driver = page.drivers.findOne({ 
  driverName: 'Grace Hopper',
  carMake: 'Nissan', 
  carModel: 'Sentra', 
  carPlate: 'FOOBAR'
});
```
Even though the first version up above is not what we'd recommend, someone could nevertheless set up their page that way, and it might be confusing that `findOne({ driverInfo: ['Grace Hopper', 'Nissan', 'Sentra', 'FOOBAR'] })` doesn't work.

Thoughts?

Reply:
I think at that point we would have to make the comparison a `deepEqual`, it would be odd to support a single level of nested comparison. I was mostly following the simple example form `lodash`, I'm unsure if they support deep equals in [find and filter](https://lodash.com/docs/4.17.4#find), but I definitely could implement it without too much trouble.

I think that this could allow users to do some interesting and maybe counter-intuitive things, so I'm split on whether it would be a good thing to add. I think I would be more comfortable with keeping it as is, without deep-comparison, because that is the more minimal API: If we add deep comparison now, we can't remove it later without making breaking changes, but we can always add it if we omit it for now.

We also have the escape hatch of allowing users to pass the predicate themselves, so they could do something like:

```js
const driver = page.drivers.findOne(driver => deepEqual(driver, { 
  driverInfo: ['Grace Hopper', 'Nissan', 'Sentra', 'FOOBAR'] 
}));
```

Reply:
As I was sanity-checking this line, the array example is the only real-world situation I could come up with where strict equality comparison wouldn't work, but I agree...if we were to do anything besides strict comparison, we would have to go with full-on `deepEqual`. 

I was mainly thinking here about predictability. Lodash `_.find` does allow something like `_.find(page.drivers.toArray(), { driverInfo: ['Grace Hopper', 'Nissan', 'Sentra', 'FOOBAR'] })`. Underscore, however, does not. `Ember.Array.findBy` does not. So...fair enough to go in either direction with this particular bit of syntactic sugar. I like your point about how we can always add this later. If there is a real need for deeper equality comparison, someone will raise it as an issue. I'm down with moving forward with strict comparison for now.